File: hooks/_trace.py
"""Optional opt-in tracing helper for oh-my-cursor hook scripts.

Activated by setting `OH_MY_CURSOR_HOOK_TRACE=1` in the environment. Each
call to :func:`trace` appends one JSON line to the trace file. The default
path is `<repo-root>/.omcs/hook-trace.log`, but it can be overridden with
`OH_MY_CURSOR_HOOK_TRACE_FILE`. The directory is created on demand.

Failures inside the tracer are swallowed so they never alter hook output.
The helper itself is read-only with respect to workflow state.
"""
from __future__ import annotations

import json
import os
from datetime import datetime, timezone
from pathlib import Path

from _repo import resolve_repo_root


_TRUE = {"1", "true", "TRUE", "yes", "on"}


def _enabled() -> bool:
    return os.environ.get("OH_MY_CURSOR_HOOK_TRACE", "") in _TRUE


def _resolve_log_path() -> Path:
    override = os.environ.get("OH_MY_CURSOR_HOOK_TRACE_FILE", "").strip()
    if override:
        return Path(override).expanduser()
    repo_root = resolve_repo_root(__file__)
    return repo_root / ".omcs" / "hook-trace.log"


def trace(record: dict) -> None:
    if not _enabled():
        return
    try:
        target = _resolve_log_path()
        target.parent.mkdir(parents=True, exist_ok=True)
        full = {
            "ts": datetime.now(timezone.utc).isoformat(),
            "pid": os.getpid(),
            **record,
        }
        with open(target, "a", encoding="utf-8") as fh:
            fh.write(json.dumps(full, ensure_ascii=False) + "\n")
    except Exception:
        return

File: hooks/_trace.ts
import * as fs from 'node:fs';
import * as path from 'node:path';
import * as process from 'node:process';
import { fileURLToPath } from 'node:url';
import { resolveRepoRoot } from './_repo.ts';

const TRUE_VALUES = new Set(['1', 'true', 'TRUE', 'yes', 'on']);

function isEnabled(): boolean {
  return TRUE_VALUES.has(process.env.OH_MY_CURSOR_HOOK_TRACE || '');
}

function resolveLogPath(): string {
  const override = (process.env.OH_MY_CURSOR_HOOK_TRACE_FILE || '').trim();
  if (override) {
    return path.resolve(override);
  }
  const currentFile = fileURLToPath(import.meta.url);
  const repoRoot = resolveRepoRoot(currentFile);
  return path.join(repoRoot, '.omcs', 'hook-trace.log');
}

export function trace(record: Record<string, any>): void {
  if (!isEnabled()) {
    return;
  }
  try {
    const target = resolveLogPath();
    const parentDir = path.dirname(target);
    fs.mkdirSync(parentDir, { recursive: true });

    const full = {
      ts: new Date().toISOString(),
      pid: process.pid,
      ...record,
    };
    fs.appendFileSync(target, JSON.stringify(full) + '\n', 'utf-8');
  } catch (err: any) {
    // Fail-silent by default; emit a warning when OMCS_TRACE_DEBUG=1 for observability
    if (process.env.OMCS_TRACE_DEBUG === '1') {
      process.stderr.write(`[omcs:trace] warn: failed to write trace log: ${err?.message ?? err}\n`);
    }
  }
}
